### PR TITLE
Add limited support for blueprint assignments

### PR DESCRIPTION
**Pull request by jackofallops**
Blocked on testing capability in CI Environment

WIP flagged until new CI environment ready.

Local tests Passing:
```
=== RUN   TestAccBlueprintAssignment_basic
--- PASS: TestAccBlueprintAssignment_basic (113.18s)
=== RUN   TestAccBlueprintAssignment_subscriptionComplete
--- PASS: TestAccBlueprintAssignment_subscriptionComplete (134.03s)
=== RUN   TestAccBlueprintAssignment_managementGroup
--- PASS: TestAccBlueprintAssignment_managementGroup (125.12s)
=== RUN   TestAccDataSourceBlueprintDefinition_basic
=== PAUSE TestAccDataSourceBlueprintDefinition_basic
=== RUN   TestAccDataSourceBlueprintDefinition_basicAtRootManagementGroup
=== PAUSE TestAccDataSourceBlueprintDefinition_basicAtRootManagementGroup
=== RUN   TestAccDataSourceBlueprintDefinition_basicAtChildManagementGroup
=== PAUSE TestAccDataSourceBlueprintDefinition_basicAtChildManagementGroup
=== RUN   TestAccDataSourceBlueprintPublishedVersion_atSubscription
=== PAUSE TestAccDataSourceBlueprintPublishedVersion_atSubscription
=== RUN   TestAccDataSourceBlueprintPublishedVersion_atRootManagementGroup
=== PAUSE TestAccDataSourceBlueprintPublishedVersion_atRootManagementGroup
=== RUN   TestAccDataSourceBlueprintPublishedVersion_atChildManagementGroup
=== PAUSE TestAccDataSourceBlueprintPublishedVersion_atChildManagementGroup
=== CONT  TestAccDataSourceBlueprintDefinition_basic
=== CONT  TestAccDataSourceBlueprintPublishedVersion_atRootManagementGroup
=== CONT  TestAccDataSourceBlueprintDefinition_basicAtChildManagementGroup
=== CONT  TestAccDataSourceBlueprintPublishedVersion_atChildManagementGroup
=== CONT  TestAccDataSourceBlueprintPublishedVersion_atSubscription
=== CONT  TestAccDataSourceBlueprintDefinition_basicAtRootManagementGroup
--- PASS: TestAccDataSourceBlueprintPublishedVersion_atChildManagementGroup (22.01s)
--- PASS: TestAccDataSourceBlueprintPublishedVersion_atRootManagementGroup (22.36s)
--- PASS: TestAccDataSourceBlueprintDefinition_basicAtChildManagementGroup (22.43s)
--- PASS: TestAccDataSourceBlueprintDefinition_basicAtRootManagementGroup (22.82s)
--- PASS: TestAccDataSourceBlueprintDefinition_basic (22.89s)
--- PASS: TestAccDataSourceBlueprintPublishedVersion_atSubscription (23.09s)
PASS
```

co-authored-by: @ArcturusZhang

### CODE REVIEW

File: azurerm/helpers/azure/identity.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package azure
```
Comment (ArcturusZhang) [inline azurerm/helpers/azure/identity.go:1]: Could we put this id parsing of a MSI into the `azurerm/internal/services/msi/parse` directory? I suppose this functionality should belong to the `msi` service, and there are quite a few resources (those support the user assigned identities) are waiting for this parsing function.

Comment (katbyte) [reply azurerm/helpers/azure/identity.go:1]: +1, we want to eventually move everything into the interal folder

File: azurerm/internal/services/blueprints/parse/blueprint_assignment.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import (
+	"fmt"
+	"strings"
+
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/azure"
+)
+
+type AssignmentId struct {
+	// "/subscriptions/00000000-0000-0000-0000-000000000000/providers/Microsoft.Blueprint/blueprintAssignments/assignSimpleBlueprint",
+	// "/managementGroups/ContosoOnlineGroup/providers/Microsoft.Blueprint/blueprintAssignments/assignSimpleBlueprint",
```
[the remark this answers is not in the document]
Comment (jackofallops) [reply azurerm/internal/services/blueprints/parse/blueprint_assignment.go:12]: The ID returned in the response is as in the comment above, without the `/providers/Microsoft.Management/` portion.  This is also the form in the swagger for the API/[examples](https://github.com/Azure/azure-rest-api-specs/blob/5f629e8a6096adf812f276263c62c574d9efe893/specification/blueprint/resource-manager/Microsoft.Blueprint/preview/2018-11-01-preview/examples/managementGroupBPAssignment/BlueprintAssignment_Create_UserAssignedMSI.json#L81), so if this is incorrect we should refer to the service team for correction?

File: azurerm/internal/services/blueprints/blueprint_definition_datasource.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+					"subscriptions",
+					"managementGroup",
+				}, false),
+			},
+
+			"scope_name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
```
[the remark this answers is not in the document]
Comment (jackofallops) [reply azurerm/internal/services/blueprints/blueprint_definition_datasource.go:43]: Revisiting this with fresh eyes after a break, I think this does need simplifying.  Since this sets the target scope, which cannot be a `managementGroup`, I'll re-work this to being a more appropriate specific item, perhaps `target_subscription_id`?

File: azurerm/internal/services/blueprints/blueprint.go [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+
+func expandArmBlueprintAssignmentResourceGroups(input string) map[string]*blueprint.ResourceGroupValue {
+	var result map[string]*blueprint.ResourceGroupValue
+	// the string has been validated by the schema, therefore the error is ignored here, since it will never happen.
+	_ = json.Unmarshal([]byte(input), &result)
+	return result
+}
+
+func expandArmBlueprintAssignmentIdentity(input []interface{}) (*blueprint.ManagedServiceIdentity, error) {
+	if len(input) == 0 {
```
Comment (mbfrahry) [inline azurerm/internal/services/blueprints/blueprint.go:135]: We'll want to confirm that input[0] isn't nil here
```suggestion
	if len(input) == 0  || input[0] == nil {
```

File: azurerm/internal/services/blueprints/blueprint_assignment_resource.go [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+		Target:  []string{string(blueprint.Succeeded)},
+		Refresh: blueprintAssignmentCreateStateRefreshFunc(ctx, client, targetScope, name),
+		Timeout: d.Timeout(schema.TimeoutCreate),
+	}
+
+	if _, err := stateConf.WaitForState(); err != nil {
+		return fmt.Errorf("failed waiting for Blueprint Assignment %q (Scope %q): %+v", name, targetScope, err)
+	}
+
+	d.SetId(*resp.ID)
```
Comment (mbfrahry) [inline azurerm/internal/services/blueprints/blueprint_assignment_resource.go:211]: We'll want to confirm that the ID isn't nil or empty before setting it

File: azurerm/internal/services/blueprints/blueprint_assignment_resource.go [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+	if resp.Identity != nil {
+		d.Set("identity", flattenArmBlueprintAssignmentIdentity(resp.Identity))
+	}
+
+	if resp.AssignmentProperties != nil {
+		if resp.AssignmentProperties.BlueprintID != nil {
+			d.Set("version_id", resp.AssignmentProperties.BlueprintID)
+		}
+
+		if resp.Parameters != nil {
```
Comment (mbfrahry) [inline azurerm/internal/services/blueprints/blueprint_assignment_resource.go:261]: It looks like this is outside of the scope of `resp.AssignmentProperties`. Thoughts on moving this out of the current if block?

File: azurerm/internal/services/blueprints/blueprint_assignment_resource.go [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+			params, err := flattenArmBlueprintAssignmentParameters(resp.Parameters)
+			if err != nil {
+				return err
+			}
+			d.Set("parameter_values", params)
+		}
+
+		if resp.ResourceGroups != nil {
```
Comment (mbfrahry) [inline azurerm/internal/services/blueprints/blueprint_assignment_resource.go:269]: It looks like this is outside of the scope of `resp.AssignmentProperties`. Thoughts on moving this out of the current if block?

File: azurerm/internal/services/blueprints/blueprint_assignment_resource.go [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+			resourceGroups, err := flattenArmBlueprintAssignmentResourceGroups(resp.ResourceGroups)
+			if err != nil {
+				return err
+			}
+			d.Set("resource_groups", resourceGroups)
+		}
+
+		// Locks
+		if locks := resp.Locks; locks != nil {
```
Comment (mbfrahry) [inline azurerm/internal/services/blueprints/blueprint_assignment_resource.go:278]: It looks like this is outside of the scope of `resp.AssignmentProperties`. Thoughts on moving this out of the current if block?

File: azurerm/internal/services/blueprints/tests/blueprint_assignment_resource_test.go [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+				Config: testAccBlueprintAssignment_rootManagementGroup(data, "testAcc_basicRootManagementGroup", "v0.1_testAcc"),
+				Check: resource.ComposeTestCheckFunc(
+					testCheckBlueprintAssignmentExists(data.ResourceName),
+				),
+			},
+			data.ImportStep(),
+		},
+	})
+}
+
```
Comment (mbfrahry) [inline azurerm/internal/services/blueprints/tests/blueprint_assignment_resource_test.go:96]: It'd be great to have a test that updates the resource as well just to confirm that all works

File: website/docs/r/blueprint_assignment.html.markdown [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+
+## Attribute Reference
+
+* `id` - the Azure Resource ID of the Blueprint Assignment
+
+* `description` - The Description on the Blueprint
+
+* `display_name` - The display name of the blueprint
+
+* `blueprint_name` - The name of the blueprint assigned
```
Comment (mbfrahry) [inline website/docs/r/blueprint_assignment.html.markdown:144]: It looks like we're missing Timeout information and how to import the resource

File: azurerm/internal/services/blueprints/blueprint_assignment_resource.go (hunk 6 above)
Comment (jackofallops) [reply azurerm/internal/services/blueprints/blueprint_assignment_resource.go:261]: This is actually inside the scope, I'd used the "shorthand" path to the object.  I'll update for the full path as `resp.AssignmentProperties.Parameters` for clarity.

File: azurerm/internal/services/blueprints/blueprint_assignment_resource.go (hunk 7 above)
Comment (jackofallops) [reply azurerm/internal/services/blueprints/blueprint_assignment_resource.go:269]: This is actually inside the scope, I'd used the "shorthand" path to the object.  I'll update for the full path as `resp.AssignmentProperties.ResourceGroups` for clarity.

File: azurerm/internal/services/blueprints/blueprint_assignment_resource.go (hunk 8 above)
Comment (jackofallops) [reply azurerm/internal/services/blueprints/blueprint_assignment_resource.go:278]: This is actually inside the scope, I'd used the "shorthand" path to the object.  I'll update for the full path as `resp.AssignmentProperties.Locks` for clarity.

File: azurerm/internal/services/blueprints/blueprint.go [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+}
+
+func blueprintAssignmentCreateStateRefreshFunc(ctx context.Context, client *blueprint.AssignmentsClient, scope, name string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		resp, err := client.Get(ctx, scope, name)
+		if err != nil {
+			return nil, "", fmt.Errorf("unable to retrieve Blueprint Assignment %q (Scope %q): %+v", name, scope, err)
+		}
+		if resp.ProvisioningState == blueprint.Failed {
+			return resp, string(resp.ProvisioningState), err
```
[the remark this answers is not in the document]
Comment (jackofallops) [reply azurerm/internal/services/blueprints/blueprint.go:70]: Good spot 👍

**Comment by jackofallops** [comment]
Tests pass:
image
